oclc_classify.py

```diff
@@ -0,0 +1,757 @@
+
+        # Add a couple extra bits of OCLC-specific information.
+        extra = {
+            OCLC.HOLDING_COUNT : edition_tag.get('holdings'),
+            OCLC.FORMAT : edition_tag.get('itemtype'),
+        }
+
+        # Get an identifier for this edition.
+        identifier, ignore = Identifier.for_foreign_id(
+            _db, Identifier.OCLC_NUMBER, oclc_number
+        )
+
+        # Create a Edition for source + identifier
+        data_source = DataSource.lookup(_db, DataSource.OCLC)
+        edition_record, new = get_one_or_create(
+            _db, Edition,
+            data_source=data_source,
+            primary_identifier=identifier,
+            create_method_kwargs=dict(
+                title=title,
+                language=language,
+                subjects=subjects,
+                extra=extra,
+            )
+        )
+
+        subjects = {}
+        for subject_type, oclc_code in (
+                (Subject.LCC, "050"),
+                (Subject.DDC, "082")):
+            classification = cls._xpath1(edition_tag,
+                "oclc:classifications/oclc:class[@tag=%s]" % oclc_code)
+            if classification is not None:
+                value = classification.get("nsfa") or classification.get('sfa')
+                identifier.classify(data_source, subject_type, value)
+
+        # Associated each contributor with the new record.
+        for author, roles in authors_and_roles:
+            edition_record.add_contributor(author, roles)
+        return edition_record, new
+
+
+class OCLCClassifyAPI(object):
+
+    BASE_URL = 'http://classify.oclc.org/classify2/Classify?'
+
+    NO_SUMMARY = '&summary=false'
+
+    def __init__(self, _db):
+        self._db = _db
+
+    @property
+    def source(self):
+        return DataSource.lookup(self._db, DataSource.OCLC)
+
+    def query_string(self, **kwargs):
+        args = dict()
+        for k, v in kwargs.items():
+            if isinstance(v, unicode):
+                v = v.encode("utf8")
+            args[k] = v
+        return urllib.urlencode(sorted(args.items()))
+
+    def lookup_by(self, **kwargs):
+        """Perform an OCLC Classify lookup."""
+        query_string = self.query_string(**kwargs)
+        url = self.BASE_URL + query_string
+        representation, cached = Representation.get(self._db, url)
+        return representation.content
+
+
+class OCLCClassifyCoverageProvider(CoverageProvider):
+    """Does title/author lookups using OCLC Classify."""
+
+    # Strips most non-alphanumerics from the title.
+    # 'Alphanumerics' includes alphanumeric characters
+    # for any language, so this shouldn't affect
+    # titles in non-Latin languages.
+    #
+    # OCLC has trouble recognizing non-alphanumerics in titles,
+    # especially colons.
+    NON_TITLE_SAFE = re.compile("[^\w\-' ]", re.UNICODE)
+
+    def __init__(self, _db, api=None, **kwargs):
+        input_identifier_types = [
+            Identifier.GUTENBERG_ID, Identifier.URI
```

Comment:
Removed Identifier.THREEM_ID from this list because the Metadata Wrangler doesn't know about 3M titles anymore.